### PR TITLE
luminous: librbd: prevent watcher from unregistering with in-flight actions

### PR DESCRIPTION
**Pull request by pdvian**
http://tracker.ceph.com/issues/23985

### CODE REVIEW

**Comment by trociny** [comment]
@dillaman Shouldn't we also backport #21141? And then this would cherry picked without conflict.

**Comment by dillaman** [comment]
@trociny It still won't apply cleanly and we don't really need that change from #21141 since that's only a race in their setup where they hang OSDs and force librados IO timeout errors in order to close the image.

File: src/librbd/Watcher.cc [hunk 1]
```diff
@@ -235,46 +240,93 @@ void Watcher::acknowledge_notify(uint64_t notify_id, uint64_t handle,
...
+
+    if (m_unregister_watch_ctx != nullptr) {
+      ldout(m_cct, 10) << "image is closing, skip rewatch" << dendl;
+      m_watch_state = WATCH_STATE_IDLE;
+      std::swap(unregister_watch_ctx, m_unregister_watch_ctx);
+    } else if (r  == -EBLACKLISTED) {
+      lderr(m_cct) << "client blacklisted" << dendl;
+    } else if (r == -ENOENT) {
+      ldout(m_cct, 5) << "object does not exist" << dendl;
+    } else if (r < 0) {
```
Comment (trociny) [inline src/librbd/Watcher.cc]: @dillaman #21141 changed `RewatchRequest` behavior to finish when any error occurs. In luminous I think this branch is dead (`RewatchRequest` will do rewatch itself).

Comment (dillaman) [reply src/librbd/Watcher.cc]: Yeah, it's unreachable but I figured no harm since the assert was dropped as part of the cherry-pick.

**Comment by yuriw** [comment]
[wip-yuri2-testing-2018-06-08-1529-luminous](https://trello.com/c/n2OD92F0/325-wip-yuri2-testing-2018-06-08-1529-luminous)

**Pull request merged**